### ISSUE TITLE
Change in a module doesn't propagate into recompilation of other module

### ISSUE THREAD

**Issue by Janiczek** [elm-lang/elm-compiler#1365]
**TL;DR:** The change in `Model` module should propagate into recompilation of `Update` module and **doesn't.**

**Slightly less TL;DR:** Compile TEA counter example with model = `Int` and everything works. Change the model type to `{thing: Int}` and the type checker doesn't find that `update` is trying to do `{thing: 0} + 1`. Runtime madness ensues (`"[object Object]1"` and so on). After deleting `elm-stuff/build-artifacts` directory the type checker works again.

----

Environment:
- elm binaries from http://install.elm-lang.org/elm-0.17-rc2.tar.gz
- Ubuntu 16.04

**EDIT:** It has been noted on #elm-dev Slack channel that this is not 0.17-only issue, it has been seen on 0.16.

----

Start with the Counter example:

```elm
# Main.elm
module Main exposing (main)

{-| The entry point to the app.
@docs main
-}

import Html.App as Html
import Model exposing (initialModel)
import Update exposing (update)
import View exposing (view)

{-| The main function. What were you expecting? -}
main : Program Never
main =
  Html.beginnerProgram
      { model = initialModel
      , update = update
      , view = view
      }
```

```elm
# Model.elm
module Model exposing (Model, Msg(..), initialModel)

type alias Model = Int
type Msg = Increment

initialModel : Model
initialModel = 0
```

```elm
# Update.elm
module Update exposing (update)

import Model exposing (Model, Msg(..))

update : Msg -> Model -> Model
update msg model =
  case msg of
    Increment ->
      model + 1
```

```elm
# View.elm
module View exposing (view)

import Html exposing (..)
import Html.Events exposing (onClick)
import Model exposing (Model, Msg(..))

view : Model -> Html Msg
view model =
  div
    []
    [ h1 [] [text (toString model)]
    , button
        [onClick Increment]
        [text "Increment"]
    ]
```

Compile with `elm make src/Main.elm --yes --output main.html`. Everything works.

----

Now change the model:

```diff
# Model.elm
module Model exposing (Model, Msg(..), initialModel)

-type alias Model = Int
+type alias Model = {thing: Int}
type Msg = Increment

initialModel : Model
-initialModel = 0
+initialModel = Model 0
```

With what's in the `Update` module, `elm-make` should complain. But running `elm make src/Main.elm --yes --output main.html` **tells us everything's okay!** This is the bug, I feel. (The app, if you try to run it now, is helplessly broken.)

----

Remove the `elm-stuff/build-artifacts` and try the long `elm make` command again, and there's the error:

```
-- TYPE MISMATCH ---------------------------------------------- ./src/Update.elm

The type annotation for `update` does not match its definition.

5│ update : Msg -> Model -> Model
            ^^^^^^^^^^^^^^^^^^^^^
The type annotation is saying:

    Msg -> { thing : Int } -> { thing : Int }

But I am inferring that the definition has this type:

    Msg -> number -> number

Detected errors in 1 module.
```

**Comment by mgold** [comment]
I think this is a dupe of [this issue](https://github.com/elm-lang/elm-make/issues/72).

**Comment by jvoigtlaender** [comment]
@mgold, in that other issue the cause of it all was that an individual module which was not the main module was recompiled in between. That does not seem to be the case here. According to @Janiczek's description, he has always been calling `elm-make` on `Main.elm`. The situation is thus different than what is reported in https://github.com/elm-lang/elm-make/issues/72.

**Comment by Janiczek** [comment]
Since yesterday me and eeue56 have tinkered with it for a bit (trying different changes in the `Model.elm` and trying different subsets of files as arguments to `elm-make`) and:

- whatever I did, I couldn't make Update.elm register that Model.elm has changed and the types are not valid
- I did however manage to get Main.elm to not compile because of Model.elm change (I changed the `thing` type from Int to String and it saw type clashes inside the `Html.beginnerProgram` function)

I can make this bug reproducibly...

**Comment by evancz** [comment]
I think @mgold is correct, so I am going to close. If it can be confirmed that it's a different problem, let's open a new one. If @Janiczek is not sure about this, I recommend asking @mgold more about it **in this issue**.
